### DISCUSSION TITLE
Proper PDF Viewer / Renderer

### DISCUSSION THREAD

**Discussion by maforget** [maforget/ComicRackCE#15] [General]
In the process of removing the old sharpPDF library (mainly use for exporting to PDF) and add one from nuget, I added PdfSharp. I've updated the part that was used to create PDF's. Then thinking I could use it to update the reader so it supports all kind of PDF, I figured out I could use that same library for it. All I could get out of it, is extracting images, like the native viewer we have. 

I then tried iText7. After updating the create the PDF part. I am again facing the same problem. I found a code on Github to render the iText objects to an image. Well the results were bad. Images had their resolution blown up (slowing down the reader) and rest wasn't rendered correctly.

There is talk about pdfium by Google and there are wrappers for it, but that has been discontinued (like everything Google).

Which brings up back to Ghostscript, which we already have, like it's the only possible way. Which is already very very slow. There are other libraries, but almost all are paid or require some kind of yearly license bullshit. I can't believe that there isn't a free PDF viewer out there.

Any idea? I've lost an entire evening rewriting twice some code to then try to figure out to update the pdf viewer just to figure out that what we have, might already be what is best.

**Comment by ghost** [comment]
Wish I knew.  I always hated Ghostscript.  It makes PDF comics look like crap.  I'll keep an eye out.

**Comment by azuravian** [comment]
There seems to be a viewer based on pdfium called PdfiumViewer.  This fork has been updated for .NET 6 and was updated as recently as 2 months ago: https://github.com/Bluegrams/PdfiumViewer

Another option is using the Community License of Syncfusion.  They do allow its use in open-source projects and have a component that offers a pdf renderer.

**Reply by maforget** [reply]
I've looked a little closer at pdfium. There are still updates to the source code. There are a couple of wrappers, the most complete is "OFF COURSE" paid. I would have liked to have the same library create the PDF and Open them. Most of the wrappers only support rendering the PDF, so I can't use it for both usage. But the most need is for rendering and it works fine for that.

So i've used a custom build of [PDFiumSharp](https://github.com/Zubastic/PDFiumSharp) library, because I don't need a reader, just being able to render a page to an image. This one has also all the Imports (with correct x86 & x64) for the native library so if needed we can add more support like creating the pdf without needing to modify the PDFiumSharp wrapper. 

The custom build I used is only to specify the location of the libraries in the Resource folder. I would like to use nuget for delivering both binaries, but I don't believe there is a way to force them to be in the Resource folder with the native nuget package and not without modifying the code generator in the PDFiumSharp library also to change the path.

For future reference: the binaries are taken from https://github.com/bblanchon/pdfium-binaries. Normally dropping in the dlls in the resource folder (and renaming them) should be enough to update the pdf library. Unless there are some breaking changes, no update to the wrapper is required.

So starting with build 51d3faec309a313fe54dde7d55a47eff85ef6124, the new pdfium library is now the default. So feedback is welcome. Please note that some images may be cached so for a fair comparison, it may be necessary to remove the existing cache when changing engines to check the difference.

There is also a new Configuration in `ComicRack.ini`, the `DisableGhostscript` was replaced by `PdfEngineToUse`, that let's you set which engine you want (Pdfium, Native or Ghostscript). Ghostscript will no longer be the default if installed unless specified in the `ComicRack.ini` file. Native is the old engine that only supported images.
